File: AngularJS/src/org/angular2/lang/html/highlighting/Angular2ColorsAndFontsPage.java
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.lang.html.highlighting;

import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighter;
import com.intellij.openapi.fileTypes.SyntaxHighlighterFactory;
import com.intellij.openapi.options.colors.AttributesDescriptor;
import com.intellij.openapi.options.colors.ColorDescriptor;
import com.intellij.openapi.options.colors.ColorSettingsPage;
import com.intellij.psi.codeStyle.DisplayPriority;
import com.intellij.psi.codeStyle.DisplayPrioritySortable;
import icons.AngularJSIcons;
import org.angular2.lang.html.Angular2HtmlFileType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.Map;

import static org.angular2.lang.html.highlighting.Angular2HtmlFileHighlighter.*;

public class Angular2ColorsAndFontsPage implements ColorSettingsPage, DisplayPrioritySortable {
  private static final AttributesDescriptor[] ATTRS;

  static {
    ATTRS = new AttributesDescriptor[]{
      new AttributesDescriptor("Two way binding `[()]`", NG_BANANA_BINDING_ATTR_NAME),
      new AttributesDescriptor("Event binding `()`", NG_EVENT_BINDING_ATTR_NAME),
      new AttributesDescriptor("Property binding `[]`", NG_PROPERTY_BINDING_ATTR_NAME),
      new AttributesDescriptor("Structural directive binding `*`", NG_TEMPLATE_BINDINGS_ATTR_NAME),
      new AttributesDescriptor("Interpolation delimiter", NG_INTERPOLATION_DELIMITER),
      new AttributesDescriptor("Plural expression", NG_EXPANSION_FORM),
      new AttributesDescriptor("Plural expression comma", NG_EXPANSION_FORM_COMMA),
      new AttributesDescriptor("Plural expression delimiter", NG_EXPANSION_FORM_DELIMITER),
      new AttributesDescriptor("Template expression", NG_EXPRESSION),
      new AttributesDescriptor("Template reference variable `#`", NG_REFERENCE_ATTR_NAME),
    };
  }

  @Override
  @NotNull
  public String getDisplayName() {
    //noinspection HardCodedStringLiteral
    return "Angular Template";
  }

  @Override
  public Icon getIcon() {
    return AngularJSIcons.Angular2;
  }

  @Override
  @NotNull
  public AttributesDescriptor[] getAttributeDescriptors() {
    return ATTRS;
  }

  @Override
  @NotNull
  public ColorDescriptor[] getColorDescriptors() {
    return ColorDescriptor.EMPTY_ARRAY;
  }

  @Override
  @NotNull
  public SyntaxHighlighter getHighlighter() {
    final SyntaxHighlighter highlighter = SyntaxHighlighterFactory.getSyntaxHighlighter(Angular2HtmlFileType.INSTANCE, null, null);
    assert highlighter != null;
    return highlighter;
  }

  @Override
  @NotNull
  public String getDemoText() {
    return "<li attr=\"value\"\n" +
           "    *ngFor=\"let hero of heroes as test\"\n" +
           "    [class.selected]=\"hero === selectedHero\"\n" +
           "    (click)=\"onSelect(hero)\">\n" +
           "  {{hero.name}}\n" +
           "</li>\n" +
           "\n" +
           "{heroes.length, plural, =0 {no heroes} =1 {one hero} other {{{heroes.length}} heroes}} found\n" +
           "\n" +
           "<input [(ngModel)]=\"hero.name\" #model=\"ngModel\"/>\n";
  }

  @Nullable
  @Override
  public Map<String, TextAttributesKey> getAdditionalHighlightingTagToDescriptorMap() {
    return null;
  }

  @Override
  public DisplayPriority getPriority() {
    return DisplayPriority.LANGUAGE_SETTINGS;
  }
}

File: AngularJS/src/org/angular2/lang/html/highlighting/Angular2ColorsAndFontsPage.kt
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.lang.html.highlighting

import com.intellij.lang.Language
import com.intellij.lang.javascript.highlighting.JSHighlighter
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighter
import com.intellij.openapi.fileTypes.SyntaxHighlighterFactory
import com.intellij.openapi.options.colors.AttributesDescriptor
import com.intellij.openapi.options.colors.ColorDescriptor
import com.intellij.openapi.options.colors.RainbowColorSettingsPage
import com.intellij.psi.codeStyle.DisplayPriority
import com.intellij.psi.codeStyle.DisplayPrioritySortable
import icons.AngularJSIcons
import org.angular2.lang.Angular2Bundle
import org.angular2.lang.expr.highlighting.Angular2HighlighterColors
import org.angular2.lang.html.Angular17HtmlFileType
import org.angular2.lang.html.Angular17HtmlLanguage
import javax.swing.Icon

class Angular2ColorsAndFontsPage : RainbowColorSettingsPage, DisplayPrioritySortable {
  override fun getDisplayName(): String {
    return Angular2Bundle.message("angular.configurable.name.angular.template")
  }

  override fun getIcon(): Icon {
    return AngularJSIcons.Angular2
  }

  override fun getAttributeDescriptors(): Array<AttributesDescriptor> {
    return ATTRS
  }

  override fun getColorDescriptors(): Array<ColorDescriptor> {
    return ColorDescriptor.EMPTY_ARRAY
  }

  override fun getHighlighter(): SyntaxHighlighter {
    return SyntaxHighlighterFactory.getSyntaxHighlighter(Angular17HtmlFileType.INSTANCE, null, null)!!
  }

  override fun getDemoText(): String {
    return """
      <li attr="value"
          *ngFor="let <ng-variable>hero</ng-variable> of <instance_variable>heroes</instance_variable> as <ng-variable>test</ng-variable>"
          [class.selected]="<ng-variable>hero</ng-variable> === <instance_variable>selectedHero</instance_variable>"
          (click)="<instance_method>onSelect</instance_method>(<ng-variable>hero</ng-variable>)">
        {{<ng-variable>hero</ng-variable>.<instance_variable>name</instance_variable>}}
      </li>
      
      @for (<ng-variable>hero</ng-variable> of <instance_variable>heroes</instance_variable>; track <instance_variable>heroes</instance_variable>.<instance_variable>name</instance_variable>) {
        {{<ng-variable>hero</ng-variable>.<instance_variable>name</instance_variable>}}
      } @empty {
        <p>No heroes!</p>
      }
      
      <div *ngIf="<ng-signal>heroSig</ng-signal>() as <ng-variable>hero</ng-variable>">{{<ng-variable>hero</ng-variable>.<instance_variable>name</instance_variable>}}</div>

      {<instance_variable>heroes</instance_variable>.<instance_variable>length</instance_variable>, plural, =0 {no heroes} =1 {one hero} other {{{<instance_variable>heroes</instance_variable>.<instance_variable>length</instance_variable>}} heroes}} found

      <input [(ngModel)]="<instance_variable>selectedHero</instance_variable>.<instance_variable>name</instance_variable>" #<ng-variable>model</ng-variable>="ngModel"/>
      
      """.trimIndent()
  }

  override fun getAdditionalHighlightingTagToDescriptorMap(): Map<String, TextAttributesKey> {
    return tags
  }

  override fun getPriority(): DisplayPriority {
    return DisplayPriority.LANGUAGE_SETTINGS
  }

  override fun isRainbowType(type: TextAttributesKey): Boolean {
    return JSHighlighter.JS_PARAMETER == type
           || JSHighlighter.JS_LOCAL_VARIABLE == type
           || JSHighlighter.JS_INSTANCE_MEMBER_VARIABLE == type
           || Angular2HighlighterColors.NG_SIGNAL == type
           || Angular2HighlighterColors.NG_VARIABLE == type
  }

  override fun getLanguage(): Language =
    Angular17HtmlLanguage.INSTANCE

  companion object {
    private val ATTRS: Array<AttributesDescriptor> = arrayOf(
      AttributesDescriptor(Angular2Bundle.message("angular.colors.banana-binding"),
                           Angular2HtmlHighlighterColors.NG_BANANA_BINDING_ATTR_NAME),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.event-binding"),
                           Angular2HtmlHighlighterColors.NG_EVENT_BINDING_ATTR_NAME),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.property-binding"),
                           Angular2HtmlHighlighterColors.NG_PROPERTY_BINDING_ATTR_NAME),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.template-binding"),
                           Angular2HtmlHighlighterColors.NG_TEMPLATE_BINDINGS_ATTR_NAME),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.signal"),
                           Angular2HighlighterColors.NG_SIGNAL),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.template-variable"),
                           Angular2HighlighterColors.NG_VARIABLE),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.interpolation-delimiter"),
                           Angular2HtmlHighlighterColors.NG_INTERPOLATION_DELIMITER),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.expansion-form"),
                           Angular2HtmlHighlighterColors.NG_EXPANSION_FORM),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.expansion-form-comma"),
                           Angular2HtmlHighlighterColors.NG_EXPANSION_FORM_COMMA),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.expansion-form-delimiter"),
                           Angular2HtmlHighlighterColors.NG_EXPANSION_FORM_DELIMITER),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.template-expression"),
                           Angular2HtmlHighlighterColors.NG_EXPRESSION),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.block-name"),
                           Angular2HtmlHighlighterColors.NG_BLOCK_NAME),
      AttributesDescriptor(Angular2Bundle.message("angular.colors.block-braces"),
                           Angular2HtmlHighlighterColors.NG_BLOCK_BRACES),
    )

    private val tags = mapOf(
      "ng-signal" to Angular2HighlighterColors.NG_SIGNAL,
      "ng-variable" to Angular2HighlighterColors.NG_VARIABLE,
      "instance_variable" to JSHighlighter.JS_INSTANCE_MEMBER_VARIABLE,
      "instance_method" to JSHighlighter.JS_INSTANCE_MEMBER_FUNCTION,
    )
  }
}